sdk/src/main/java/ch/admin/bag/covidcertificate/eval/verification/CertificateVerifier.kt

```diff
@@ -68,32 +84,96 @@ internal class CertificateVerifier(private val nationalRulesVerifier: NationalRu
+			val type = dccHolder.certType ?: return@withContext CheckSignatureState.INVALID(ErrorCodes.SIGNATURE_TYPE_INVALID)
+
+			val timestampError = TimestampService.decode(dccHolder)
+			if (timestampError != null) {
+				return@withContext CheckSignatureState.INVALID(timestampError)
+			}
+
+			// Repeat decode chain to get and verify COSE signature
+			val encoded = PrefixIdentifierService.decode(dccHolder.qrCodeData)
+				?: return@withContext CheckSignatureState.INVALID(ErrorCodes.DECODE_PREFIX)
+			val compressed = Base45Service.decode(encoded)
+				?: return@withContext CheckSignatureState.INVALID(ErrorCodes.DECODE_BASE_45)
+			val cose = DecompressionService.decode(compressed)
+				?: return@withContext CheckSignatureState.INVALID(ErrorCodes.DECODE_Z_LIB)
+
+			val valid = VerificationCoseService.decode(signatures.certs, cose)
+			if (valid) {
+				CheckSignatureState.SUCCESS
+			} else {
+				CheckSignatureState.INVALID(ErrorCodes.SIGNATURE_COSE_INVALID)
+			}
 		} catch (e: Exception) {
 			CheckSignatureState.ERROR(Error(ErrorCodes.SIGNATURE_UNKNOWN, e.message.toString(), dccHolder))
 		}
 	}
 
+	/**
+	 * Checks whether a certificate has been revoked
+	 *
+	 * @param dccHolder The object returned from the decoder
+	 * @param revokedCertificates The list of revoked certificate identifiers from the trust list
+	 * @return Outcome state of the revocation check
+	 */
 	private suspend fun checkRevocationStatus(
 		dccHolder: DccHolder,
 		revokedCertificates: RevokedCertificates
 	) = withContext(Dispatchers.Default) {
+		// Revocation is not possible for light certificates, so this check always returns SUCCESS for them
+		if (dccHolder.isLightCertificate()) return@withContext CheckRevocationState.SUCCESS
+
 		try {
-			Eval.checkRevocationStatus(dccHolder, revokedCertificates)
+			val revokedCertificateService = RevokedHealthCertService(revokedCertificates)
+			val containsRevokedCertificate = revokedCertificateService.isRevoked(dccHolder.euDGC!!)
+
+			if (containsRevokedCertificate) {
+				CheckRevocationState.INVALID(ErrorCodes.REVOCATION_REVOKED)
+			} else {
+				CheckRevocationState.SUCCESS
+			}
 		} catch (e: Exception) {
 			CheckRevocationState.ERROR(Error(ErrorCodes.REVOCATION_UNKNOWN, e.message.toString(), dccHolder))
 		}
 	}
 
+	/**
+	 * Checks whether a certificate conforms to a set of national rules
+	 *
+	 * @param dccHolder The object returned from the decoder
+	 * @param nationalRulesVerifier The verifier responsible to verify a certificate against a national rule set
+	 * @param ruleSet The national rule set from the trust list
+	 * @return Outcome state of the national rules check
+	 */
 	private suspend fun checkNationalRules(
 		dccHolder: DccHolder,
 		nationalRulesVerifier: NationalRulesVerifier,
 		ruleSet: RuleSet
 	) = withContext(Dispatchers.Default) {
 		try {
-			Eval.checkNationalRules(dccHolder, nationalRulesVerifier, ruleSet)
+			if (dccHolder.isLightCertificate()) {
```

Comment:
This sounds a bit like we want the DccHolder to either hold a EU or a light certificate.
If this is the case, maybe we could check this in the DccHolder's constructor/init?
On the other hand, would there be a case in the wallet app where we want a DccHolder to hold both at the same time?

Reply:
We always keeps the EU Cert. The light is optional in my opinion, if it exits we show it.

Reply:
Okay, so in other words: the EuDcc is always there, but the DccLight takes higher priority (both in display and in validation).

Reply:
No, that's not the case. The DccHolder is a holder object of a single certificate. While it's true that we always store the EU cert and optionally a cert light, the DccHolder is the decoded QR code data, which is either an EU cert or a cert light.
Otherwise the certificate decoding mechanism has to receive two input strings, one for the EU QR Code and one for the QR Code light.